dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp

```diff
@@ -134,52 +133,72 @@ void ReadBufferFromKafkaConsumer::unsubscribe()
     consumer->unsubscribe();
 }
 
-/// Do commit messages implicitly after we processed the previous batch.
-bool ReadBufferFromKafkaConsumer::nextImpl()
+bool ReadBufferFromKafkaConsumer::reset()
 {
-    /// NOTE: ReadBuffer was implemented with an immutable underlying contents in mind.
-    ///       If we failed to poll any message once - don't try again.
-    ///       Otherwise, the |poll_timeout| expectations get flawn.
-    if (stalled || stopped || !allowed)
-        return false;
-
-    if (current == messages.end())
+    if (current != messages.end())
     {
-        if (intermediate_commit)
-            commit();
+        ++current;
+        return true;
+    }
+
+    return poll(0);
```

Comment:
why 0?

Reply:
It's optimisation. We don't want to wait inside `IBlockStream::read()` for full polling timeout - so we try to poll instantly if anything is still available. If not, then we fail to reset and stop reading.

Reply:
It's not well documented, and digging the source is time-consuming. Are you sure that 0 means "return immediately" not "there is not timeout"?

Reply:
And actually it would be better to split ISource refactoring part (preserving the existing logic) which will do exactly what in PR title - i.e. "Incapsulate resetting feature inside ISource" and the part which actually changes the Kafka behaviour in 2 separate PR.

Reply:
Also i think that this 0 here can lead to insertion of tiny blocks. I.e. if you will have a topic with 1 msg per 500ms, you will do one row inserts 2 times per sec (i.e. braking the contract of max_block_size / max_execution_time / poll timeout).

Reply:
Can we just preserve existing poll logic to avoid unneeded side effects?